### PR TITLE
RecursionError in integrate fixed

### LINKED ISSUE #21034: (Integration) regressions?
Hello, I am trying to integrate the following expressions (`f`) which work under SymPy v1.4 but fail on both the master version and 1.7.1.

Behavior in v1.4 is:

```
>>> from sympy import *
>>> x = Symbol('x', real=True, nonzero=True)

>>> f1 = x*(-x**4/asin(5)**4 - x*sinh(x + log(asin(5))) + 5)
>>> integrate(f1, x)
-x**6/(6*asin(5)**4) - x**2*cosh(x + log(asin(5))) + 5*x**2/2 + 2*x*sinh(x + log(asin(5))) - 2*cosh(x + log(asin(5)))

>>> f2 = (x + cosh(cos(4)))/(x*(x + 1/(12*x)))
>>> integrate(f2, x)
log(x**2 + 1/12)/2 + 2*sqrt(3)*cosh(cos(4))*atan(2*sqrt(3)*x)
```

For `f1`: v1.7.1 and master raise the following error: `RecursionError: maximum recursion depth exceeded`
For `f2`: v1.7.1 raises CoercionFailed error (issue possibly related to #21024), while master fails with `RecursionError: maximum recursion depth exceeded`

### PR DESCRIPTION
**Pull request by nisargsc**
#### References to other Issues or PRs

Partial fix for #21034 

#### Brief description of what is fixed or changed
Applies the diff mentions by @oscarbenjamin [here](https://github.com/sympy/sympy/issues/21034#issuecomment-790178650). This fixes one of the problems in the issue, i.e. `RecurstionError` caused due to `not x.is_extended_real` in `sympy/core/expr.py`

#### Release Notes

* core
  * Fixed a bug in `Expr.round` that could lead to infinite recursion in `integrate`.

### CODE REVIEW

File: sympy/core/tests/test_arit.py [hunk 1]
```diff
@@ -2321,3 +2323,16 @@ def test_issue_18507():
...
+
+
+def test_issue_21034():
+    x = Symbol('x', real=True, nonzero=True)
+    f1 = x*(-x**4/asin(5)**4 - x*sinh(x + log(asin(5))) + 5)
+    e = -I*log((re(asin(5)) + I*im(asin(5)))/sqrt(re(asin(5))**2 + im(asin(5))**2))
+    e2 = -I*log((re(asin(5)) + I*im(asin(5)))/sqrt(re(asin(5))**2 + im(asin(5))**2))/pi
+    assert integrate(f1, x) == \
+        -x**6/(6*asin(5)**4) - x**2*cosh(x + log(asin(5))) + 5*x**2/2 + 2*x*sinh(x + log(asin(5))) - 2*cosh(x + log(asin(5)))
+    assert f1.is_zero is None
```
Comment (oscarbenjamin) [inline sympy/core/tests/test_arit.py:2335]: Looking at this again there's really just one bug here which is this:
```
expr = -I*log((re(asin(5)) + I*im(asin(5)))/sqrt(re(asin(5))**2 + im(asin(5))**2))/pi
expr.round() # hangs
```
We can just add a test for that here in this file. The integral test should go in `sympy.integrals` somewhere.

Comment (Nisarg-Chaudhari) [reply sympy/core/tests/test_arit.py:2335]: on it. 
also in `sympy.integrals` I'm adding it to `sympy/interals/test_integrals.py` is it fine?

**Comment by oscarbenjamin** [comment]
Does this also fix the other integral from the issue?

**Comment by Nisarg-Chaudhari** [comment]
No, the other one still gives `RecursionError: maximum recursion depth exceeded`.

**Comment by oscarbenjamin** [comment]
Okay, since it doesn't fully fix the issue I've changed the OP to say "partial fix". If the OP says "fixes #12345" then when it is merged the issue will be closed but we don't want to close the issue unless it is fully fixed.

I've also edited the release not to make it clear that this was a bug in `Expr.round`.

This looks good so I'll merge it now. Thanks!

**Pull request merged**